mqtt/mqtt-bridge/src/client.rs

```diff
@@ -0,0 +1,227 @@
+
+        Box::pin(async move {
+            let expiry = Utc::now()
+                + chrono::Duration::from_std(DEFAULT_TOKEN_DURATION_SECS).unwrap_or_else(|_| {
+                    chrono::Duration::minutes(DEFAULT_TOKEN_DURATION_MINS.try_into().unwrap())
+                });
+            let mut password = None;
+            if let Some(ts) = token_source {
+                password = match ts.get(&expiry).await {
+                    Ok(x) => Some(x),
+                    Err(e) => {
+                        error!("Failed to get token for connection {} {}", address, e);
+                        None
+                    }
+                }
+            };
+
+            let io = TcpStream::connect(address).await;
+            io.map(|io| (io, password))
+        })
+    }
+}
+
+pub struct MqttClient<T>
+where
+    T: EventHandler,
+{
+    client: Client<TcpConnection<SasTokenSource>>,
+    event_handler: T,
+}
+
+impl<T: EventHandler> MqttClient<T> {
+    pub fn new(
+        address: &str,
+        keep_alive: Duration,
+        clean_session: bool,
+        event_handler: T,
+        connection_credentials: &Credentials,
+    ) -> Self {
+        let (client_id, token_source) = match connection_credentials {
+            Credentials::Provider(provider_settings) => (
+                provider_settings.device_id().into(),
+                Some(SasTokenSource::new(connection_credentials.clone())),
+            ),
+            Credentials::PlainText(creds) => (
+                creds.client_id().into(),
+                Some(SasTokenSource::new(connection_credentials.clone())),
+            ),
+            Credentials::Anonymous(client_id) => (client_id.into(), None),
+        };
+
+        let client = if clean_session {
+            Client::new(
+                Some(client_id),
+                None,
+                None,
+                TcpConnection {
+                    address: address.into(),
+                    token_source,
+                },
+                DEFAULT_MAX_RECONNECT,
+                keep_alive,
+            )
+        } else {
+            Client::from_state(
+                client_id,
+                None,
+                None,
+                TcpConnection {
+                    address: address.into(),
+                    token_source,
+                },
+                DEFAULT_MAX_RECONNECT,
+                keep_alive,
+            )
+        };
+
+        MqttClient {
+            client,
+            event_handler,
+        }
+    }
+
+    pub fn shutdown_handle(&self) -> Result<ShutdownHandle, ShutdownError> {
+        self.client
+            .shutdown_handle()
+            .map_or(Err(ShutdownError::ClientDoesNotExist), |shutdown_handle| {
+                Ok(ShutdownHandle(shutdown_handle))
+            })
+    }
+
+    pub async fn handle_events(mut self) -> Result<(), ClientConnectError> {
+        while let Some(event) = self.client.try_next().await.unwrap_or_else(|e| {
```

Comment:
I thought Err here means that we should stop polling stream because of IO error. Could you double check that we can continue work with a client in this case?

Reply:
you're right, client can't be reused so it needs to be recreated. As discussed offline, I will have a different PR to handle this case.